### ISSUE TITLE
Generate configuration file

### ISSUE THREAD

**Issue by zachriggle** [pwndbg/pwndbg#137]
Pwndbg has grown to have sufficiently many configuration points that perhaps it's a good idea to have a command to generate a configuration file.

The config file should be valid GDB syntax, and sets the **current** values, while showing the default values.

This would expose all of the options to new and current users in an easy-to-manage fashion.

**Comment by anthraxx** [comment]
neat feature :smiley: :heart:

**Comment by disconnect3d** [comment]
Where should one put the generated `configfile`?

Either I do something wrong or something is bugged.

Suppose I have a config:
```
pwndbg> config
Name                       Value (Def) Documentation
----------------------------------------------------
auto-save-search           False       automatically pass --save to "search" command
```

The `configfile` command says:
```
pwndbg> configfile
# auto-save-search: automatically pass --save to "search" command
# default: False
py pwndbg.config.auto_save_search=False
```

So I put the last three lines in my `.gdbinit` just after `source pwndbg/gdbinit.py`.

Then I restart gdb and the `configfile` command doesn't show up the `auto_save_search` configuration.

**Comment by zachriggle** [comment]
In fairness, I never actually tested the config files, I just assumed they'd work as expected.

Reopened for investigation <3

**Comment by disconnect3d** [comment]
@zachriggle: Fixed in 1f56e5e7a64019dea793271c6953f4ff99393059 - the problem was that we changed the `pwndbg.config.attribute` so it was no more a `pwndbg.config.Parameter` instance and so checks like `isinstance(v, pwndbg.config.Parameter)` failed.

The proposed solution fixes this with generating settings such as:
```
py pwndbg.config.ida_rpc_host.value = '192.168.0.123'
```
Instead of:
```
py pwndbg.config.ida_rpc_host='192.168.0.123'
```

(The `.value` has been added).

I am also filtering the values and generating only the changed ones. If there are no changed values additional information is shown. See the screen below.
![pwndbg_configfile](https://cloud.githubusercontent.com/assets/10009354/20868285/5b06bbd8-ba58-11e6-802e-20a5e4d75043.png)


I have also added some additional information to the `config` and `theme` commands:
![pwndbg_config](https://cloud.githubusercontent.com/assets/10009354/20868273/eb88a2d0-ba57-11e6-9104-11ae9f953460.png)


PS: I have also fired code refactoring & imports optimization from my PyCharm. It is based on pep8 style guide. If you don't like it I can revert those refactor/imports optimization changes (however I would strongly encourage to be valid with pep8, it might make the project a little bit more attractive for new developers).

**Comment by zachriggle** [comment]
Do I need to cherry-pick that commit, or is it already fixed?

Regarding import style, we already enforce it with `isort`.  Feel free to make any changes to that config file you want.  I prefer a CLI versus PyCharm, and just want things to be consistent -- don't really care pep8 vs whatever.

**Comment by disconnect3d** [comment]
Not really, as I wrote in #139, the configuration is still broken. Let me recall.

Basically, when having a .gdbinit such as:

```
source ~/pwndbg/gdbinit.py
py pwndbg.config.ida_rpc_host.value = '192.168.0.123'
```

The parameter is changed but ... after the initialization. So during the initialization, the parameters have their default values.

My first attempt was to change the .gdbinit to this:
```
python
pwndbg_config = {
    'ida_rpc_host': '192.168.0.123',
    # other config parameters
}
end

source ~/pwndbg/gdbinit.py
```
and then fetch the proper value from proper dict (either `pwndbg_config` or `pwndbg_theme`) when creating parameter instances (if those values are there).

But it doesn't work since the `pwndbg_config` variable is accessible only from `pwndbg/gdbinit.py` (and not from `pwndbg/__init__.py` or other modules imported by the start script). There is a possibility to create variable global across modules, but [probably it is too hackish](http://stackoverflow.com/questions/142545/python-how-to-make-a-cross-module-variable).

I will just change `configfile` and `themefile` to fill `user_config.py` or `user_theme.py` files which will be empty by default (or having a comment that would explain their existence).

**Comment by disconnect3d** [comment]
The "issue" described above (https://github.com/pwndbg/pwndbg/issues/137#issuecomment-264654345) has been revisited in  PR #139 - it ended up with f728a2870a2b22373702cdd81649f1b7b4be9b80 which:
* adds more output into `config`/`theme`/`configfile`/`themefile` commands (e.g. where to put the generated commands)
* adds info about commands into FEATURES.md
* changes the config syntax to use gdb native syntax (`set <var> <value>`) instead of python invocations (`py pwndbg.config.<paramname>.value = <python-value>`)



@zachriggle  I guess we can close it now ;)

**Comment by zachriggle** [comment]
Closed via #139
